### PR TITLE
DAOS-13938 dfuse: adjust offset in readdir cache entry list

### PR DESCRIPTION
**Pull request by johannlombardi**
The dfuse readdir caching feature stats entries returned by dfs readdir that might be legimitely deleted by another process or another node. When this happens, the dfuse code decides to skip the deleted entry, but might leave the readdir entry cache in an inconsistent state causing the following assertion failure in dfuse_dre_drop():
drc->drc_next_offset == expected_offset + 1 || drc->drc_next_offset == READDIR_EOD

This patch fixes the offset in the linked list of cached readdir entries to support skipped entries.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/client/dfuse/ops/readdir.c [hunk 1]
```diff
@@ -759,6 +761,12 @@ dfuse_do_readdir(struct dfuse_info *dfuse_info, fuse_req_t req, struct dfuse_obj
 			}
 
 			if (drc) {
+				if (oh->doh_rd_nextc != NULL)
```
Comment (jolivier23) [inline src/client/dfuse/ops/readdir.c:764]: minor nit, any time I have comments, I use braces on if statement and checkpatch never cared.   I personally hate the no braces for single line rule style but that is another story.

Comment (johannlombardi) [reply src/client/dfuse/ops/readdir.c:764]: I am with you. I will fix once i need to refresh the patch. Probably after @ashleypittman's review.

Comment (ashleypittman) [reply src/client/dfuse/ops/readdir.c:764]: I prefer the brevity of no braces for single-line-if statements as they're often simple but I completely agree that if there are comments involved it needs to be clear what code block they belong to.

**Review by ashleypittman — COMMENTED** [review]
This is a complex issue and I need time to come up to speed with the code but it looks exactly the area I'd expect to need updating so I think Johann is on the right track here.

The reproducer for this issue needs two clients, one performing readdir and another modifying the filesystem, the bug relates to a crash in dfuse when performing the readdir request, this will want testing with both ENOENT errors which should be ignored and other errors which should be (probably) propagated to the readdir call.

I'm working through the code and looking at a reproducer and performing a better review of this PR currently.

**Review by ashleypittman — APPROVED** [review]
This looks correct but I'm not sure it's sufficient to solve the problem entirely.

There are two places that dfs_lookup/create_entry are called in this code and we need to handle both ENOENT and other errors from both sites so at the very least significantly more validation is required here.

**Comment by johannlombardi** [comment]
let's land, i will fix the other in a different PR.

**Pull request merged**